test/general/test_entrance_rando.py

```diff
@@ -0,0 +1,376 @@
+
+        self.assertIsNone(e.connected_region)
+        self.assertEqual([], r1.entrances)
+
+        self.assertEqual(1, len(r1.exits))
+        self.assertEqual(e, r1.exits[0])
+
+        self.assertEqual(1, len(r2.entrances))
+        self.assertIsNone(r2.entrances[0].parent_region)
+        self.assertEqual("r2", r2.entrances[0].name)
+        self.assertEqual(EntranceType.ONE_WAY, r2.entrances[0].randomization_type)
+        self.assertEqual(2, r2.entrances[0].randomization_group)
+
+
+class TestRandomizeEntrances(unittest.TestCase):
+    def test_determinism(self):
+        """tests that the same output is produced for the same input"""
+        multiworld1 = generate_test_multiworld()
+        multiworld1.worlds[1].random = multiworld1.per_slot_randoms[1]
+        generate_disconnected_region_grid(multiworld1, 5)
+        multiworld2 = generate_test_multiworld()
+        multiworld2.worlds[1].random = multiworld2.per_slot_randoms[1]
+        generate_disconnected_region_grid(multiworld2, 5)
+
+        result1 = randomize_entrances(multiworld1.worlds[1], False, directionally_matched_group_lookup)
+        result2 = randomize_entrances(multiworld2.worlds[1], False, directionally_matched_group_lookup)
+        self.assertEqual(result1.pairings, result2.pairings)
+        for e1, e2 in zip(result1.placements, result2.placements):
+            self.assertEqual(e1.name, e2.name)
+            self.assertEqual(e1.parent_region.name, e1.parent_region.name)
+            self.assertEqual(e1.connected_region.name, e2.connected_region.name)
+
+    def test_all_entrances_placed(self):
+        """tests that all entrances and exits were placed, all regions are connected, and no dangling edges exist"""
+        multiworld = generate_test_multiworld()
+        multiworld.worlds[1].random = multiworld.per_slot_randoms[1]
+        generate_disconnected_region_grid(multiworld, 5)
+
+        result = randomize_entrances(multiworld.worlds[1], False, directionally_matched_group_lookup)
+
+        self.assertEqual([], [entrance for region in multiworld.get_regions()
+                              for entrance in region.entrances if not entrance.parent_region])
+        self.assertEqual([], [exit_ for region in multiworld.get_regions()
+                              for exit_ in region.exits if not exit_.connected_region])
+        # 5x5 grid + menu
+        self.assertEqual(26, len(result.placed_regions))
+        self.assertEqual(80, len(result.pairings))
+        self.assertEqual(80, len(result.placements))
+
+    def test_coupling(self):
+        """tests that in coupled mode, all 2 way transitions have an inverse"""
+        multiworld = generate_test_multiworld()
+        multiworld.worlds[1].random = multiworld.per_slot_randoms[1]
+        generate_disconnected_region_grid(multiworld, 5)
+        seen_placement_count = 0
+
+        def verify_coupled(state: ERPlacementState, placed_entrances: List[Entrance]):
```

Comment:
Could use `_: ERPlacementState` to signal that the variable is unused. My IDE is happier when I do this, anyway